// @microsoft/sp-webpart-base/lib-commonjs/common/Enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedHostType = exports.SPClientSideComponentType = void 0;
/**
 * @internal
 * These should be kept in sync with the same enum defined in STS/stsom
 */
var SPClientSideComponentType;
(function (SPClientSideComponentType) {
    // Unknown client-side component type
    SPClientSideComponentType[SPClientSideComponentType["Unknown"] = 0] = "Unknown";
    // Client-side web part; a drop-in control that is part of the end user page authoring experience
    SPClientSideComponentType[SPClientSideComponentType["WebPart"] = 1] = "WebPart";
    // Client-side application; the architectural component that manages the appearance and behavior of an entire web page
    SPClientSideComponentType[SPClientSideComponentType["Application"] = 2] = "Application";
    // Client-side library; a library containing reusable JavaScript code and/or resources
    SPClientSideComponentType[SPClientSideComponentType["Library"] = 3] = "Library";
    // Client-side extension; a plug-in component that customizes or augments the functionality of a client-side application
    SPClientSideComponentType[SPClientSideComponentType["Extension"] = 4] = "Extension";
    // Client-side provider hosted app; this is a webpart containing an app hosted elsewhere
    SPClientSideComponentType[SPClientSideComponentType["ProviderHostedApp"] = 5] = "ProviderHostedApp";
    // Client-side assembly; this is the assembly that manages an application
    SPClientSideComponentType[SPClientSideComponentType["Assembly"] = 6] = "Assembly";
    // Client-side Adaptive Card Extension
    SPClientSideComponentType[SPClientSideComponentType["AdaptiveCardExtension"] = 7] = "AdaptiveCardExtension";
})(SPClientSideComponentType = exports.SPClientSideComponentType || (exports.SPClientSideComponentType = {}));
/**
 * @internal
 * These should be kept in sync with the same enum defined in STS/stsom
 */
var SupportedHostType;
(function (SupportedHostType) {
    SupportedHostType[SupportedHostType["TeamsTab"] = 0] = "TeamsTab";
    SupportedHostType[SupportedHostType["SharePointFullPage"] = 1] = "SharePointFullPage";
    SupportedHostType[SupportedHostType["SharePointWebPart"] = 2] = "SharePointWebPart";
    SupportedHostType[SupportedHostType["Dashboard"] = 3] = "Dashboard";
})(SupportedHostType = exports.SupportedHostType || (exports.SupportedHostType = {}));
