// @microsoft/sp-webpart-base/lib-commonjs/common/KillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHandleAsyncPropertyUpdateKSActivated = exports.isClearPreviousStateWhenSetWebPartDataKSActive = exports.isImageRefactorKSActivated = exports.isAsyncCompleteKSActivated = exports.isMeasurePerformanceImpactOfWebPartFlickerKSActivated = exports.isWebPartManifestListCacheRefreshKSActivated = exports.isAccessibleRootKSActivated = exports.isDashboardWebPartMediumCardBreakPointKSActivated = exports.isAddMoreRenderErrorInfoKSActivated = exports.isDashboardWebPartVerticalLayoutCardWidthFix = exports.isBotDrivenACEManifestLoadLogicKSActivated = exports.isExemptWebPartWithDynamicDataFromLazyLoadKSActivated = exports.isAvoidMutatingWebPartDataBeforeDeserializationKSActive = exports.isAlwaysShowIsolatedACEOnTeamsKSActive = exports.isVivaHomeIsolatedAceCheckKSActive = exports.isPropertyPaneUpdateASHAVetoKSActivated = exports.isIsolatedACEWPRenderAsyncKSActivated = exports.isPreventLateIframeRedirectKSActivated = exports.isVersionHeaderKillSwitchActivated = exports.isUberIsolatedACEFlightKSActivated = exports.isSafariIsolationKSActivated = exports.isLogWebPartLoadErrorKSActivated = exports.isFixMissingContextKSActivated = exports.isDynamicIframedPropertyPaneHandlingKSActivated = exports.isIsolatedAdaptiveCardExtensionIframedPropertyPaneKSActivated = exports.isChainDisposeCallsAfterOninitKSActivated = exports.isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated = exports.isIsolatedWebPartTabStopKSActivated = exports.fixACERequiringWPinSupportedHostsKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
var KillSwitches = /** @class */ (function () {
    function KillSwitches() {
    }
    // THIS KILLSWITCH HAS BEEN ACTIVATED
    KillSwitches.isTeamsContextKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated(sp_core_library_1.Guid.parse('a5daffb0-652e-4ee1-a638-e290089e1de7'), '09/11/2020', 'Refactor a method of IWebPartHost as async, as part of PR to enable teams context.');
    };
    return KillSwitches;
}());
exports.default = KillSwitches;
function fixACERequiringWPinSupportedHostsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6c94e4d7-b951-48c7-b282-318ad7525e96'
    /* '10/5/2021', 'Temporary fix to stop requiring ACEs have SharePointWebPart in their supportedHosts array' */
    );
}
exports.fixACERequiringWPinSupportedHostsKSActivated = fixACERequiringWPinSupportedHostsKSActivated;
function isIsolatedWebPartTabStopKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d8618918-ac55-4979-b9d4-be944a4feb95'
    /* '10/25/2021', 'adds tabindex=0 to the isolated webpart iframe' */
    );
}
exports.isIsolatedWebPartTabStopKSActivated = isIsolatedWebPartTabStopKSActivated;
function isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b8c50ded-d048-4eb9-bc41-21190d9d3bbb'
    /* '01/20/2022', 'Work around to enable property pane for isolated ACE web' */
    );
}
exports.isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated = isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated;
function isChainDisposeCallsAfterOninitKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ca984138-b518-4cf1-b201-2efe9322bb5e'
    /* '02/01/2022', 'Ensures that calls to dispose always wait for any activing onInit calls' */
    );
}
exports.isChainDisposeCallsAfterOninitKSActivated = isChainDisposeCallsAfterOninitKSActivated;
function isIsolatedAdaptiveCardExtensionIframedPropertyPaneKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ab6a720f-9915-4635-9b30-71b5d83e0647'
    /* '01/21/2022', 'Creates a functional iframed property pane for isolated ACEs' */
    );
}
exports.isIsolatedAdaptiveCardExtensionIframedPropertyPaneKSActivated = isIsolatedAdaptiveCardExtensionIframedPropertyPaneKSActivated;
function isDynamicIframedPropertyPaneHandlingKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1a2da4d0-9678-4a79-95a4-f0b9a7732dc2'
    /* '02/02/2022', 'Dynamically creates and assigns an iframed property pane to selected IBaseFramedWebParts' */
    );
}
exports.isDynamicIframedPropertyPaneHandlingKSActivated = isDynamicIframedPropertyPaneHandlingKSActivated;
function isFixMissingContextKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e1e44790-a263-4f43-8b40-5b5e9b170300'
    /* '02/22/2022', 'Fix missing context when passing down to propertypane' */
    );
}
exports.isFixMissingContextKSActivated = isFixMissingContextKSActivated;
function isLogWebPartLoadErrorKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c041bc0e-fcb5-4a3e-a4bc-d65a00a39a86'
    /* '02/24/2022', 'Check if exception thrown during loadWebPart is an error and ' */
    );
}
exports.isLogWebPartLoadErrorKSActivated = isLogWebPartLoadErrorKSActivated;
function isSafariIsolationKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('61a9578e-c6d2-4605-970f-222a8b29196c' /* '03/13/22', 'Skip rendering isolated ACEs in Safari' */);
}
exports.isSafariIsolationKSActivated = isSafariIsolationKSActivated;
function isUberIsolatedACEFlightKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('07684c68-2e8e-4a25-b7e0-330781a92661'
    /* '03/16/22', 'If the KS is not pulled check the uber IsolatedAdaptiveCardExtensionWeb flight otherwise check APEXIsolatedACEWEB flights' */
    );
}
exports.isUberIsolatedACEFlightKSActivated = isUberIsolatedACEFlightKSActivated;
function isVersionHeaderKillSwitchActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8a6c4749-5c36-4214-aa06-1ea308d679e2'
    /* '5/17/22', 'Send odsp-web version to service worker' */
    );
}
exports.isVersionHeaderKillSwitchActivated = isVersionHeaderKillSwitchActivated;
function isPreventLateIframeRedirectKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('56151a3e-a100-403c-9e91-cb5d1677ed66'
    /* '06/13/2022', 'instantiate IframeLoadRedirector earlier and support pre-emptively preventing redirects' */
    );
}
exports.isPreventLateIframeRedirectKSActivated = isPreventLateIframeRedirectKSActivated;
function isIsolatedACEWPRenderAsyncKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e8edc935-7cd4-463b-81f5-aa1cce74d7d0'
    /* '06/27/2022', 'Captures if async web part has finished rendering using LKG or updated content.' */
    );
}
exports.isIsolatedACEWPRenderAsyncKSActivated = isIsolatedACEWPRenderAsyncKSActivated;
function isPropertyPaneUpdateASHAVetoKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('21f00b7b-3ed6-4c3a-bd49-8639460de07e'
    /* '08/02/2022', 'Add monitor to capture property pane update scenario for ASHA' */
    );
}
exports.isPropertyPaneUpdateASHAVetoKSActivated = isPropertyPaneUpdateASHAVetoKSActivated;
function isVivaHomeIsolatedAceCheckKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('f993c5d6-5637-47bc-802d-6da6dda11a27' /* '08/24/22', 'Fix the isolated ACE flight check for Viva Home' */);
}
exports.isVivaHomeIsolatedAceCheckKSActive = isVivaHomeIsolatedAceCheckKSActive;
function isAlwaysShowIsolatedACEOnTeamsKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('6c457592-8c32-472e-b340-349c014d6bc2' /* '10/20/2022', 'Always show isolated cards on Teams' */);
}
exports.isAlwaysShowIsolatedACEOnTeamsKSActive = isAlwaysShowIsolatedACEOnTeamsKSActive;
function isAvoidMutatingWebPartDataBeforeDeserializationKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('34173619-d841-476a-98c3-af5a79007103'
    /* '12/07/2022', 'Avoid mutating web part data when calling _internalSetWebPartData' */
    );
}
exports.isAvoidMutatingWebPartDataBeforeDeserializationKSActive = isAvoidMutatingWebPartDataBeforeDeserializationKSActive;
function isExemptWebPartWithDynamicDataFromLazyLoadKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bbd78e03-ff18-40ed-9a94-973fc659366c' /* '12/01/2022', 'Update getWebPartsExemptedFromViewPortLoading to check for containsDynamicDataSource' */);
}
exports.isExemptWebPartWithDynamicDataFromLazyLoadKSActivated = isExemptWebPartWithDynamicDataFromLazyLoadKSActivated;
function isBotDrivenACEManifestLoadLogicKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6bc17fec-be41-4460-8f40-b31e293e3d91' /* '1/03/2023', 'Update manifest loading logic for Bot-Driven ACEs' */);
}
exports.isBotDrivenACEManifestLoadLogicKSActivated = isBotDrivenACEManifestLoadLogicKSActivated;
function isDashboardWebPartVerticalLayoutCardWidthFix() {
    return sp_core_library_1._SPKillSwitch.isActivated('43ce53d4-53da-48c4-acbd-20ff31eff5ee' /* '1/09/2023', 'Fix medium card width for dashboard web part in vertical layout.' */);
}
exports.isDashboardWebPartVerticalLayoutCardWidthFix = isDashboardWebPartVerticalLayoutCardWidthFix;
function isAddMoreRenderErrorInfoKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('290c0d46-6ee0-4d0b-aee5-ef3f01e4020e' /* '01/11/2023', 'Add more RenderError Information' */);
}
exports.isAddMoreRenderErrorInfoKSActivated = isAddMoreRenderErrorInfoKSActivated;
function isDashboardWebPartMediumCardBreakPointKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8b372850-6897-4ef2-aecd-cd615019e544' /* '03/06/2023', 'Reduce size of medium cards if we cannot fit at least two min width medium cards in a row for dashboard web part.' */);
}
exports.isDashboardWebPartMediumCardBreakPointKSActivated = isDashboardWebPartMediumCardBreakPointKSActivated;
function isAccessibleRootKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('33558902-7587-4229-a570-633976fa332b' /* '03/28/2023', 'Change the DOM hierarchy of where the accessible web part label is rendered' */);
}
exports.isAccessibleRootKSActivated = isAccessibleRootKSActivated;
function isWebPartManifestListCacheRefreshKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('98bb76c9-49c4-4736-80b9-88ff41d65cac' /* '04/24/2023', 'Refreshes the webpart cache list after loading the new manifest list' */);
}
exports.isWebPartManifestListCacheRefreshKSActivated = isWebPartManifestListCacheRefreshKSActivated;
function isMeasurePerformanceImpactOfWebPartFlickerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('abb86541-760a-44b4-88f4-57a01af6e8c2' /* '05/09/2023', 'Measure performance impact of web part flicker' */);
}
exports.isMeasurePerformanceImpactOfWebPartFlickerKSActivated = isMeasurePerformanceImpactOfWebPartFlickerKSActivated;
function isAsyncCompleteKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6e0a6d0e-ec0c-4667-85ce-198b4bbe78af' /* '05/17/2023', 'Allow render complete async callback to work if called quickly' */);
}
exports.isAsyncCompleteKSActivated = isAsyncCompleteKSActivated;
function isImageRefactorKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('138ce21a-eef4-4e30-b350-4225400c72ac'
    /* 05/30/2023, 'Add KS to disable new image webpart experience' */
    );
}
exports.isImageRefactorKSActivated = isImageRefactorKSActivated;
function isClearPreviousStateWhenSetWebPartDataKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('7b14ee6c-0ab8-44b2-89a7-597a504254e0'
    /* '06/15/2023', 'Clear _previousState when setWebPart to avoid unnecessary dirty bit' */
    );
}
exports.isClearPreviousStateWhenSetWebPartDataKSActive = isClearPreviousStateWhenSetWebPartDataKSActive;
function isHandleAsyncPropertyUpdateKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bcd05704-f61a-473d-ba66-39f0bfa0891d'
    /* '07/05/2023', 'Allows asynchronous updating of properties and setting of dirty bit' */
    );
}
exports.isHandleAsyncPropertyUpdateKSActivated = isHandleAsyncPropertyUpdateKSActivated;
