// @fluentui/utilities/lib-amd/shadowDom/contexts/MergeStylesDefaultContext.js
define(["require", "exports", "@fluentui/merge-styles", "react"], function (require, exports, merge_styles_1, React) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.MergeStylesDefaultContext = exports.getNewContext = void 0;
    var noop = function () { return false; };
    var noopShadow = function () { return merge_styles_1.DEFAULT_SHADOW_CONFIG; };
    var noopRootStylesheets = function () { return new Map(); };
    var noopUndefined = function () { return undefined; };
    var getNewContext = function () {
        return {
            stylesheets: new Map(),
            useAdoptedStylesheetEx: noop,
            useAdoptedStylesheet: noop,
            useShadowConfig: noopShadow,
            useMergeStylesShadowRootContext: noopUndefined,
            useHasMergeStylesShadowRootContext: noop,
            useMergeStylesRootStylesheets: noopRootStylesheets,
            useWindow: noopUndefined,
            useStyled: noopUndefined,
        };
    };
    exports.getNewContext = getNewContext;
    exports.MergeStylesDefaultContext = React.createContext((0, exports.getNewContext)());
});
