// @microsoft/sp-webpart-base/lib-commonjs/chunks/IframedWebPartController/ReservedIframeQueryParamKeys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReservedIframeQueryParamKeys = void 0;
/**
 * Enum which contains the reserved query string keys which are used by the
 * framework while constructing the iframe url.
 */
var ReservedIframeQueryParamKeys;
(function (ReservedIframeQueryParamKeys) {
    ReservedIframeQueryParamKeys["List"] = "list";
    ReservedIframeQueryParamKeys["Id"] = "id";
    ReservedIframeQueryParamKeys["WebPartInstanceId"] = "webPartInstanceId";
    ReservedIframeQueryParamKeys["NewWebPartInstanceId"] = "newWebPartInstanceId";
    ReservedIframeQueryParamKeys["ParentWindowOrigin"] = "parentWindowOrigin";
    ReservedIframeQueryParamKeys["ComponentId"] = "componentId";
    ReservedIframeQueryParamKeys["HostedInCanvas"] = "hostedInCanvas";
    ReservedIframeQueryParamKeys["HostedInTeams"] = "hostedInTeams";
    ReservedIframeQueryParamKeys["OpenPropertyPane"] = "openPropertyPane";
    ReservedIframeQueryParamKeys["OpenDetailsPropertyPane"] = "openDetailsPropertyPane";
})(ReservedIframeQueryParamKeys = exports.ReservedIframeQueryParamKeys || (exports.ReservedIframeQueryParamKeys = {}));
